common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -795,10 +801,27 @@ public void registerExecutor(String appId, ExecutorShuffleInfo executorInfo) {
   }
 
   /**
-   * Close the DB during shutdown
+   * Shutdown mergedShuffleCleaner and close the DB during shutdown
    */
   @Override
   public void close() {
+    if (!mergedShuffleCleaner.isShutdown()) {
+      // SPARK-40186：Use two phases shutdown refer to
+      // https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html
+      // Use two phases shutdown can prevent new tasks and wait for executing tasks to
+      // complete gracefully, and once timeout is reached, we want to interrupt running tasks,
+      // so that they fail. This is to prevent updates to shuffle state db after it is closed.
+      try {
+        mergedShuffleCleaner.shutdown();
+        // Wait a while for existing tasks to terminate
+        if (!mergedShuffleCleaner.awaitTermination(cleanerShutdownTimeout, TimeUnit.SECONDS)) {
+          shutdownMergedShuffleCleanerNow();
+        }
+      } catch (InterruptedException ignored) {
+        shutdownMergedShuffleCleanerNow();
+        Thread.currentThread().interrupt();
```

Comment:
Why do we need to call interrupt()?

Reply:
for preserve interrupt status, refer to the demo in https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html

<img width="591" alt="image" src="https://user-images.githubusercontent.com/1475305/188651981-003bec4c-7404-448f-9d53-aecae6e9e587.png">

Reply:
Ok..but why do we need to preserve the interrupt status?

Reply:
@Ngone51 Do you mean in this specific case why we are preserving ? Or in general ?
I cant think of a good reason to do it for this specific case (other than the fact that it is idiomatic for this sort of usage :) ) since we are already in NM shutdown at this time

Reply:
Yes, this is a textbook style.

After `Thread.currentThread().interrupt()` is called, the current thread can respond to the interrupt again.

But for this scenario, it should have no practical effect due to `db.close()` method does not respond to interrupts

Reply:
Thanks for the explanation. @mridulm @LuciferYang